robocorp-code/src/robocorp_code/inspector/image/_image_inspector.py

```diff
@@ -0,0 +1,135 @@
+import threading
+import typing
+from typing import Any, Callable, Optional, List
+
+from robocorp_ls_core.protocols import IEndPoint
+from robocorp_ls_core.callbacks import Callback
+from robocorp_ls_core.robotframework_log import get_logger
+
+from ._bridge import ImageBridge
+
+log = get_logger(__name__)
+
+
+class _PickerThread(threading.Thread):
+    def __init__(
+        self,
+        bridge: Optional[ImageBridge],
+        confidence: Optional[int],
+        on_pick,
+    ) -> None:
+        threading.Thread.__init__(self)
+
+        self._stop_event = threading.Event()
+        self.bridge = bridge
+        self.on_pick = on_pick
+        self.confidence = confidence
+
+    def run(self) -> None:
+        try:
+            self._run()
+        except Exception:
+            import traceback
+
+            traceback.print_exc()
+
+    def _run(self) -> None:
+        assert self.bridge is not None
+
+        result_dict = self.bridge.pick(confidence=self.confidence)
+        self.on_pick(result_dict)
+
+
+class _ValidateThread(threading.Thread):
+    def __init__(
+        self,
+        bridge: Optional[ImageBridge],
+        image_base64: str,
+        confidence: Optional[int],
+        on_validate,
+    ) -> None:
+        threading.Thread.__init__(self)
+
+        self._stop_event = threading.Event()
+        self.bridge = bridge
+        self.image_base64 = image_base64
+        self.confidence = confidence
+        self.on_validate = on_validate
+
+    def run(self) -> None:
+        try:
+            self._run()
+        except Exception:
+            import traceback
+
+            traceback.print_exc()
+
+    def _run(self) -> None:
+        assert self.bridge is not None
+
+        matches = self.bridge.validate(
+            image_base64=self.image_base64, confidence=self.confidence
+        )
+        self.on_validate(matches)
+
+
+class IOnPickCallback(typing.Protocol):
+    def __call__(self, locator_info_tree: List[dict]):
+        pass
+
+    def register(self, callback: Callable[[Any], Any]) -> None:
+        pass
+
+    def unregister(self, callback: Callable[[Any], Any]) -> None:
+        pass
+
+
+class ImageInspector:
+    def __init__(self, endpoint: Optional[IEndPoint] = None) -> None:
+        """
+        Args:
+            endpoint: If given notifications on the state will be given.
+        """
+        # callbacks
+        self.on_pick: IOnPickCallback = Callback()
+        self.on_validate: IOnPickCallback = Callback()
+        # internals
+        self._image_bridge = ImageBridge(endpoint=endpoint, logger=log)
+        # threads
+        self._picker_thread: Optional[_PickerThread] = None
+        self._validate_thread: Optional[_ValidateThread] = None
+
+    def start_pick(self, confidence: Optional[int] = None) -> None:
+        log.info(">>> Image - Start pick...")
+
+        self._picker_thread = _PickerThread(
```

Comment:
This class spawns threads but I don't think it handles the internal state properly at this point.

i.e.:

If there's an existing `self._picker_thread` I guess it should first stop that one and then start a new one and the `on_pick` should only be related to the current thread.

Also, I'm not sure if this is meant to be thread-safe or not. Given that this is handling a bunch of internal state, can you do the following:

If it must be thread-safe, you probably should use a `threading.RLock` for checking existing and stopping it / creating new version/starting it in this case (and any access to `self._picker_thread`).

If it doesn't need to be thread safe because you know you won't access things from multiple threads verify that only the same threads access the function calls here.

i.e.: something as `self._use_in_thread = threading.current_thread()` and then at the start of each method do something as `assert self._use_in_thread == threading.current_thread()`